airsonic-main/src/main/java/org/airsonic/player/dao/LegacyHsqlDaoHelper.java

```diff
@@ -31,29 +51,226 @@ public void checkpoint() {
+
+    /**
+     * Perform a backup of the HSQLDB database, to a timestamped directory.
+     * @return the path to the backup directory
+     */
+    private Path performHsqldbDatabaseBackup() throws IOException {
+
+        String timestamp = new SimpleDateFormat("yyyyMMddHHmmss").format(new Date());
+        Path source = Paths.get(settingsService.getDefaultJDBCPath()).getParent();
+        Path destination = Paths.get(String.format("%s.backup.%s", settingsService.getDefaultJDBCPath(), timestamp));
+
+        LOG.info("Backing up HSQLDB database to {}...", destination.toString());
+        try (Stream<Path> stream = Files.walk(source)) {
+
+            List<Path> sources = stream.collect(toList());
+            List<Path> destinations = sources.stream()
+                    .map(source::relativize)
+                    .map(destination::resolve)
+                    .collect(toList());
+
+            for (int i = 0; i < sources.size(); i++) {
+                LOG.info("Copying {} to {}", sources.get(i), destinations.get(i));
+                Path currentSource = sources.get(i);
+                Path currentDestination = destinations.get(i);
+                if (currentDestination.toFile().isDirectory()) {
+                    currentDestination.toFile().mkdirs();
+                } else {
+                    currentDestination.getParent().toFile().mkdirs();
+                    Files.copy(sources.get(i), destinations.get(i));
+                }
+            }
+        }
+
+        LOG.info("HSQLDB database backup complete.");
+        return destination;
+    }
+
+    /**
+     * Perform an in-place database upgrade from HSQLDB 1.x to 2.x.
+     */
+    private void performHsqldbDatabaseUpgrade() throws SQLException {
+        // This will upgrade HSQLDB from 1.8 to 2.5 on the first connection.
+        try (Connection conn = DriverManager.getConnection(settingsService.getDefaultJDBCUrl())) {
+            LOG.info("Database connection established. Current version is: {}.", conn.getMetaData().getDatabaseProductVersion());
+            // On upgrade, the official documentation recommends that we
+            // run 'SHUTDOWN SCRIPT' to compact all the database into a
+            // single SQL file.
+            //
+            // In practice, if we don't do that, we did not observe issues
+            // immediately but after the upgrade.
+            LOG.info("Shutting down database...");
+            try (Statement st = conn.createStatement()) {
+                st.execute("SHUTDOWN SCRIPT");
+            }
+        }
+
+        LOG.info("Database connection closed. Current version is: {}.", getHsqldbDatabaseVersion());
+    }
+
+    /**
+     * Checks if the database is correctly shut down
+     */
+    private boolean isHsqldbDatabaseCorrectlyShutdown() {
+
+        // The db/airsonic.data file may or may not exist (contains CACHED tables).
+        // The db/airsonic.backup file may or may not exist (not used during normal operation).
```

Comment:
Is this true after `SHUTDOWN SCRIPT`?
I think only the script and the properties file remains.

In fact, I thought this is how SHUTDOWN SCRIPT is different from SHUTDOWN COMPACT?
The former compresses down to script files, the latter additionally generates the backup and data files?

Reply:
After `SHUTDOWN SCRIPT` only script and properties remain, you're right. But I'm also calling this before the upgrade to check the assumptions I've made on the state of the db, so it needs to handle un-upgraded databases too.